File: site/src/lib/polls.ts
// 投票配置（单一真相源 = polls.json，前端与 API 共用同一份定义）。
// 加新投票只需往 polls.json 加一项；票表 / API / widget 不动。
import pollsData from '../data/polls.json';
import type { Locale } from './data';

export interface PollOption {
  key: string;
  mid?: number; // 数值中点（单选折算“大众平均%”用）
  label: Record<string, string>;
}
export interface Poll {
  code: string;
  type: 'single' | 'slider';
  scope: string;
  q: Record<string, string>;
  conclusion?: Record<string, string>;
  options: PollOption[];
}

export const POLLS: Poll[] = (pollsData as any).polls as Poll[];

// 取本地化文案：缺失回退 en（站点只维护中英文母本）
export function locText(m: Record<string, string>, locale: Locale): string {
  return m[locale] ?? m.en ?? Object.values(m)[0] ?? '';
}

// 单选聚合折算大众平均分（用 option.mid 加权），无 mid 或无票返回 null
export function avgFromCounts(poll: Poll, counts: Record<string, number>): number | null {
  let n = 0, s = 0;
  for (const o of poll.options) {
    if (o.mid == null) return null;
    const c = counts[o.key] || 0;
    n += c; s += c * o.mid;
  }
  return n ? Math.round(s / n) : null;
}

export function totalCount(counts: Record<string, number>): number {
  return Object.values(counts || {}).reduce((a, b) => a + (b || 0), 0);
}

// 各选项百分比：{optionKey: pct(0-100 整数)}
export function pctOf(counts: Record<string, number>): Record<string, number> {
  const total = totalCount(counts);
  const out: Record<string, number> = {};
  for (const k of Object.keys(counts || {})) out[k] = total ? Math.round((counts[k] * 100) / total) : 0;
  return out;
}

// 模板填充：{name} {n} {avg} 以及 {pct.<optionKey>}。
// 只替换传入的键，未提供的占位符原样保留（便于分两步填：先服务端填 name，再客户端填 n/avg/pct）。
export function fillTpl(
  tpl: string,
  vars: { name?: string; n?: number | string; avg?: number | string | null },
  pct?: Record<string, number>,
): string {
  let s = tpl || '';
  if (pct) s = s.replace(/\{pct\.(\w+)\}/g, (m, k) => (k in pct ? String(pct[k]) : m));
  s = s.replace(/\{(\w+)\}/g, (m, k) => (k in vars ? ((vars as any)[k] == null ? '' : String((vars as any)[k])) : m));
  return s;
}

File: aijobrisk-go/internal/polls/polls.go
// Package polls 投票定义与纯函数助手（单一真相源 = data/polls.json，前端组件与 API 共用）。
package polls

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
)

type Option struct {
	Key   string            `json:"key"`
	Mid   *float64          `json:"mid"`
	Label map[string]string `json:"label"`
}
type Poll struct {
	Code       string            `json:"code"`
	Type       string            `json:"type"`
	Scope      string            `json:"scope"`
	Q          map[string]string `json:"q"`
	Conclusion map[string]string `json:"conclusion"`
	Options    []Option          `json:"options"`
}

var (
	Polls  []Poll
	ByCode = map[string]Poll{}
)

// Load 从 dir/polls.json 载入投票定义。
func Load(dir string) error {
	b, err := os.ReadFile(filepath.Join(dir, "polls.json"))
	if err != nil {
		return err
	}
	var doc struct {
		Polls []Poll `json:"polls"`
	}
	if err := json.Unmarshal(b, &doc); err != nil {
		return err
	}
	Polls = doc.Polls
	for _, p := range Polls {
		ByCode[p.Code] = p
	}
	return nil
}

// OptionKeys 某投票的合法选项键集合。
func OptionKeys(code string) map[string]bool {
	out := map[string]bool{}
	if p, ok := ByCode[code]; ok {
		for _, o := range p.Options {
			out[o.Key] = true
		}
	}
	return out
}

// LocText 取本地化文案（缺失回退 en，再回退首个）。
func LocText(m map[string]string, locale string) string {
	if v, ok := m[locale]; ok {
		return v
	}
	if v, ok := m["en"]; ok {
		return v
	}
	for _, v := range m {
		return v
	}
	return ""
}

// TotalCount 计票合计。
func TotalCount(counts map[string]int) int {
	t := 0
	for _, c := range counts {
		t += c
	}
	return t
}

// PctOf 各选项百分比（0-100 整数）。
func PctOf(counts map[string]int) map[string]int {
	total := TotalCount(counts)
	out := map[string]int{}
	for k, c := range counts {
		if total > 0 {
			out[k] = int(float64(c)*100/float64(total) + 0.5)
		} else {
			out[k] = 0
		}
	}
	return out
}

// AvgFromCounts 用 option.mid 加权折算大众平均分；任一选项无 mid 或无票返回 nil。
func AvgFromCounts(p Poll, counts map[string]int) *int {
	n, s := 0, 0.0
	for _, o := range p.Options {
		if o.Mid == nil {
			return nil
		}
		c := counts[o.Key]
		n += c
		s += float64(c) * (*o.Mid)
	}
	if n == 0 {
		return nil
	}
	v := int(s/float64(n) + 0.5)
	return &v
}

var (
	rePct  = regexp.MustCompile(`\{pct\.(\w+)\}`)
	reVar  = regexp.MustCompile(`\{(\w+)\}`)
)

// FillTpl 模板填充：{pct.<key>} 先填（若给 pct），再填 {name}/{n}/{avg}；未提供的占位符原样保留。
func FillTpl(tpl string, vars map[string]string, pct map[string]int) string {
	s := tpl
	if pct != nil {
		s = rePct.ReplaceAllStringFunc(s, func(m string) string {
			k := rePct.FindStringSubmatch(m)[1]
			if v, ok := pct[k]; ok {
				return strconv.Itoa(v)
			}
			return m
		})
	}
	s = reVar.ReplaceAllStringFunc(s, func(m string) string {
		k := reVar.FindStringSubmatch(m)[1]
		if v, ok := vars[k]; ok {
			return v
		}
		return m
	})
	return s
}
